rules/epmark.go

```diff
@@ -0,0 +1,141 @@
+	log "github.com/sirupsen/logrus"
+)
+
+// Endpoint Mark Mapper (EPM) provides set of functions to manage allocation/free endpoint mark bit
+// given a mark bit mask.
+type EndpointMarkMapper interface {
+	GetEndpointMark(ep string) (uint32, error)
+	RemoveEndpointMark(ep string)
+	SetEndpointMark(ep string, mark uint32) (error)
+}
+
+type DefaultEPMarkManager struct {
+	markBitsManager *markbits.MarkBitsManager
+	maxPosition     int
+
+	activeEndpointToPosition map[string]int
+	activePositionToEndpoint map[int]string
+
+	mutex sync.Mutex
+}
+
+func NewEndpointMarkMapper(markMask uint32) EndpointMarkMapper {
+	markBitsManager := markbits.NewMarkBitsManager(markMask, "endpoint-iptable-mark")
+
+	return &DefaultEPMarkManager{
+		markBitsManager:          markBitsManager,
+		maxPosition:              markBitsManager.CurrentFreeNumberOfMark(), // This includes zero
+		activeEndpointToPosition: map[string]int{},
+		activePositionToEndpoint: map[int]string{},
+	}
+}
+
+func (epmm *DefaultEPMarkManager) GetEndpointMark(ep string) (uint32, error) {
+	length := len(ep)
+	if length == 0 {
+		return 0, errors.New("Invalid endpoint name")
+	}
+
+	epmm.mutex.Lock()
+	defer epmm.mutex.Unlock()
+
+	// Return current mark for Endpoint if it already has one.
+	if pos, ok := epmm.activeEndpointToPosition[ep]; ok {
+		mark, err := epmm.markBitsManager.MapNumberToMark(pos)
+		if err != nil {
+			return 0, err
+		}
+		return mark, nil
+	}
+
+	// Try to allocate a position based on a simple hash from endpoint name.
+	// Make sure we are likely to get a good performance based on average 10 to 50 pods per node.
+
+	// Take last two bytes or one byte from name and take the modulus of max position.
+	var total int
+	bytes := []byte(ep)
+	if length >= 2 {
+		total = int(uint32(bytes[length-2])<<8 + uint32(bytes[length-1]))
+	} else {
+		total = int(bytes[length-1])
+	}
+	prospect := total % epmm.maxPosition //Get hash position prospect
+	if prospect == 0 {
+		// Make sure it is not zero
+		prospect++
+	}
+
+	if _, alloced := epmm.activePositionToEndpoint[prospect]; alloced {
+		// We got a collision, prospect position has been allocated. Walk through and find an empty position.
+		for i := 0; i < epmm.maxPosition && alloced; i++ {
+			prospect++
+			if prospect >= epmm.maxPosition {
+				prospect = 1 // Make sure it is not zero
+			}
+
+			_, alloced = epmm.activePositionToEndpoint[prospect]
+		}
+
+		if alloced {
+			return 0, errors.New("No mark position left")
+		}
+	}
+
+	epmm.activePositionToEndpoint[prospect] = ep
+	epmm.activeEndpointToPosition[ep] = prospect
+	mark, err := epmm.markBitsManager.MapNumberToMark(prospect)
+	if err != nil {
+		return 0, err
+	}
+	log.WithFields(log.Fields{
+		"ep":   ep,
+		"Pos":  prospect,
+		"Mark": mark,
+	}).Debug("Got endpoint mark")
+	return mark, nil
+}
+
+func (epmm *DefaultEPMarkManager) RemoveEndpointMark(ep string) {
+	epmm.mutex.Lock()
+	defer epmm.mutex.Unlock()
+
+	if _, ok := epmm.activeEndpointToPosition[ep]; ok {
```

Comment:
You can do `pos, ok := ...` here and remove the next line since pos will be in scope inside the if block.  (That works even if the element isn't present.)

Reply:
Yes. Done.